### PR TITLE
alternative callback-builder-style notation for actionsMap

### PR DESCRIPTION
**Pull request by phryneas**
This is what I had suggested on twitter. Right now, it's just implemented on createReducer, but if you agree that we go forward with this, I'll add it to extraReducers, too.

### CODE REVIEW

File: type-tests/files/createReducer.typetest.ts [hunk 1]
```diff
@@ -63,3 +63,55 @@ function expectType<T>(p: T) {}
...
+      expectType<{ type: 'increment' }>(action)
+      // typings:expect-error
+      expectType<{ type: 'decrement' }>(action)
+      // typings:expect-error - this cannot be inferred and has to be manually specified
+      expectType<{ type: 'increment'; payload: number }>(action)
+    })
+
+    builder.addCase(
+      increment,
+      (state, action: ReturnType<typeof increment>) => state
```
Comment (markerikson) [inline type-tests/files/createReducer.typetest.ts]: Somehow I'd actually kind of figure this action type could get inferred automatically, but maybe that's not the case.

**Comment by markerikson** [comment]
Looks pretty neat.  I assume we wouldn't want to allow this for `createSlice`'s `reducers` arg, because the point of that is we're using that to _generate_ the action types, right?  This is only good for when we've already got action types defined and are now generating reducers to match them.

File: type-tests/files/createReducer.typetest.ts (hunk 1 above)
Comment (phryneas) [reply type-tests/files/createReducer.typetest.ts]: It does (see line 86), this is just there to test if it can actually also be specified without error.

**Comment by phryneas** [comment]
> Looks pretty neat. I assume we wouldn't want to allow this for `createSlice`'s `reducers` arg, because the point of that is we're using that to _generate_ the action types, right? This is only good for when we've already got action types defined and are now generating reducers to match them.

Exactly. I'm seeing a use case for `createReducer`'s `actionMap`, and `createSlice`s `extraReducers`.

Maybe also for some kind of async middleware in the future.

**Comment by markerikson** [comment]
Yeah, go ahead and rig it up for `extraReducers`, then.

> Maybe also for some kind of async middleware in the future.

How so?

**Comment by phryneas** [comment]
Intersting sidenote: there's no codesandbox being generated for this. Maybe because it's a Draft PR? 🤔

**Comment by phryneas** [comment]
> Yeah, go ahead and rig it up for `extraReducers`, then.
> 
> > Maybe also for some kind of async middleware in the future.
> 
> How so?

The most simple async middleware I could think of would be something like an actionEffectsMap, either as an argument on `configureStore`, or even on a per-slice-level. Something like

```typescript
configureStore({
// ...
asyncEffects: {
  [todosRequested.type]:  async (action: ReturnType<typeof todosRequested>, { getState, dispatch }) => {
    // async stuff here
  } 
}
});
```

If we did something like that, that could benefit from a similar alternative notation.

**Comment by markerikson** [comment]
Hey, good point.

And yes, a simple action -> callback map is basically what I'm picturing.  Already opened up a threat to discuss it over in #237 .

**Comment by markerikson** [comment]
btw, after this, it'd be nice if we could add an actual "Usage with TypeScript" docs page that rephrases the info in the "Advanced Tutorial" and talks about using this builder.

**Comment by phryneas** [comment]
Yeah. Also, the docblocks will need some updating. Do you know how to correctly write docblocks for overloaded methods?

**Comment by markerikson** [comment]
Not specifically, no, and tbh the types I've listed in the API reference pages are sorta hand-wavy anyway (like references to a `ReducerFunction` type).

I'd say write the page for the callback builder first, then update the other couple pages to point to that and say that the args can either be a hand-written object or the result of the builder.

**Comment by phryneas** [comment]
Code-wise this should be it.

I'm gonna take a look at the docs tomorrow - I should get some sleep now :)

**Comment by markerikson** [comment]
@phryneas : maybe we lost the CodeSandbox integration due to the repo name change?

**Comment by phryneas** [comment]
> @phryneas : maybe we lost the CodeSandbox integration due to the repo name change?

I guess so. Could you remove & re-add the github integration for a test?

I've now added the missing docblock comments.

As for the documentation page, writing just doesn't come to me today.
But I've created a codeSandbox with basically the content I would add to that page, without all the fancy words inbetween. If you're in a writing mood, could you take over from there?
https://codesandbox.io/s/rtk-ts-documentation-z8phk

Otherwise, I'll try again in a few days. That writing mood is unpredictable.

**Comment by markerikson** [comment]
I didn't even set up the actual integration - I filled out the form and got an email from someone, I think.

Lemme see if I can find that email and reply asking for an update.

I'm trying to convince myself to work on a blogpost atm, so I'm gonna hold off on trying to write docs for this.

**Comment by markerikson** [comment]
I just created a new issue for the TS usage page and copied these comments over there.

**Comment by phryneas** [comment]
> I didn't even set up the actual integration - I filled out the form and got an email from someone, I think.

I think you have to go to https://github.com/apps/codesandbox and re-activate it for the renamed sandbox.

Aside from that - is this ready in your eyes/can I merge this or do you want to wait for the documentation to catch up?

**Comment by garethx** [comment]
You'll need to enable the new repo in the CodeSandbox CI app - we only see reduxjs/redux-starting-kit enabled atm.

**Comment by markerikson** [comment]
Yeah, we really oughta merge this in :)

Can you add info on this to the API docs as well?  We can do that as a separate PR.

**Comment by phryneas** [comment]
@garethx by now, codesandbox builds are running again, but I just noticed that codeSandbox cannot find types on the CI builds. See this sandbox: https://codesandbox.io/s/friendly-hoover-g0sni
The same code in a non-CI-build seems to work: https://codesandbox.io/s/relaxed-bhabha-iqp0u
Is this a known issue? If not, where would be the best place to open an issue? The codesandbox repo structure is a bit confusing to me.

**Pull request merged**